zcash_client_sqlite/src/wallet/init.rs

```diff
@@ -19,6 +25,240 @@ use {
+                ufvk TEXT NOT NULL,
+                address TEXT,
+                transparent_address TEXT
+            );",
+        )?;
+
+        let mut stmt_fetch_accounts =
+            transaction.prepare("SELECT account, address FROM accounts")?;
+
+        let rows = stmt_fetch_accounts.query_map(NO_PARAMS, |row| {
+            let account: u32 = row.get(0)?;
+            let address: String = row.get(1)?;
+            Ok((AccountId::from(account), address))
+        })?;
+
+        for row in rows {
+            let (account, address) = row?;
+            let decoded_address =
+                decode_payment_address(self.params.hrp_sapling_payment_address(), &address)?
+                    .ok_or_else(|| {
+                        SqliteClientError::CorruptedData(format!(
+                            "Not a valid Sapling payment address: {}",
+                            address
+                        ))
+                    })?;
+
+            let usk = UnifiedSpendingKey::from_seed(&self.params, &self.seed, account).unwrap();
+            let ufvk = usk.to_unified_full_viewing_key();
+
+            if let Some(dfvk) = ufvk.sapling() {
+                let (_, expected_address) = dfvk.default_address();
+                if expected_address != decoded_address {
+                    return Err(SqliteClientError::CorruptedData("Decoded Sapling address does not match the default Sapling address derived from the provided seed.".to_string()));
+                }
+            }
+
+            add_account_internal(&self.params, transaction, "accounts_new", account, &ufvk)?;
+        }
+        transaction.execute_batch(
+            "DROP TABLE accounts;
+            ALTER TABLE accounts_new RENAME TO accounts;",
+        )?;
+
+        //
+        // Update the sent_notes table to inclue an output_pool column that
+        // is respected by the uniqueness constraint
+        //
+
+        transaction.execute_batch(
+            "CREATE TABLE IF NOT EXISTS sent_notes_new (
+                id_note INTEGER PRIMARY KEY,
+                tx INTEGER NOT NULL,
+                output_pool INTEGER NOT NULL ,
+                output_index INTEGER NOT NULL,
+                from_account INTEGER NOT NULL,
+                address TEXT NOT NULL,
+                value INTEGER NOT NULL,
+                memo BLOB,
+                FOREIGN KEY (tx) REFERENCES transactions(id_tx),
+                FOREIGN KEY (from_account) REFERENCES accounts(account),
+                CONSTRAINT tx_output UNIQUE (tx, output_pool, output_index)
+            );
+            INSERT INTO sent_notes_new 
+                (id_note, tx, output_pool, output_index, from_account, address, value, memo)
+                SELECT id_note, tx, 2, output_index, from_account, address, value, memo 
+                FROM sent_notes;
+            DROP TABLE sent_notes;
+            ALTER TABLE sent_notes_new RENAME TO sent_notes;
+            PRAGMA foreign_keys = ON;",
```

Comment:
If we don't make it all the way here, does the database potentially end up in a weird state where foreign keys are turned off?

Is there a kind of try-finally we can use to ensure that foreign keys are turned back on regardless of whether the migration succeeds?

I imagine a potential recovery mode if migration fails is to drop the tables, recreate the tables, then rescan the wallet since all data should be recoverable given time+seed.  If we ever do some kind of recovery, it would be nice to ensure that foreign keys aren't turned of by accident here.

Reply:
Everything is in the scope of a single SQLite transaction; I'll need to check to see whether SQLite DDL operations are transactional.

Reply:
Note that scoping the pragma more narrowly:
```sql
PRAGMA foreign_keys = ON;

CREATE TABLE sent_notes_new [...];
INSERT INTO sent_notes_new [...];

PRAGMA foreign_keys = OFF;
DROP TABLE sent_notes;
ALTER TABLE sent_notes_new RENAME TO sent_notes;
PRAGMA foreign_keys = ON;
```
has a different effect because it enforces the foreign key constraints while `sent_notes_new` is being created. (Setting the pragma to ON does not check existing rows added when it was OFF, I think.) I believe those constraints should hold for `sent_notes_new`, so this may be preferable.

If this is changed then I would suggest doing the same for `accounts_new`, although that has no foreign key constraints.

Reply:
Ah, okay, I understand now. Fixed.